Fix lifecycle bug when stopping NeoDataSource

davidegrohmann opened the pull request:
If the checkpointing step fails in NeoDataSource.stop(), we never call
life.shutdown(). Hence we'll never unmap the store files from the page
cache. So later when we try to close the page cache it will fail due
to still mapped files.  This change we'll make sure that we always
shutdown all the lifecycled instances in the internal life in
NeoDataSource

davidegrohmann committed:
Simplify close in CommonAbstractStore

tinwelint commented on community/kernel/src/main/java/org/neo4j/kernel/impl/store/CommonAbstractStore.java:
```diff
@@ -851,18 +851,6 @@ protected void assertNotClosed()
-        {
-            try
-            {
-                closeStoreFile();
```
This is something of an unneccesary change in 3.0, also it's already done in 3.1

davidegrohmann replied:
Last time I checked, it was in 3.1 as well.

davidegrohmann replied:
Anyhow, what's wrong in doing this change in 3.0 as well?

MishaDemianenko replied:
no reason to touch it in a maintenance branch

davidegrohmann replied:
I see no reason to not remove stupid code in an old branch. Being super religious about this stuff is not very helpful.

tinwelint replied:
I think it was removed today or something. Funny how two devs notice the same thing at the same time after years just sitting there in the code base :) I'm sure it's not a problem to remove this in 3.0 as well though

MishaDemianenko replied:
you can freely do it in an active brunch, no reason to touch something else there in general just because you can, that's why it's called maintenance

tinwelint commented on community/kernel/src/main/java/org/neo4j/kernel/NeoStoreDataSource.java:
```diff
@@ -876,19 +874,21 @@ public synchronized void stop()
-            if ( databaseHealth.isHealthy() )
+            // We cannot throw here since we need to shutdown without exceptions,
+            // so let's make the checkpointing part of the life, so LifeSupport can handle exceptions properly
+            life.add( new LifecycleAdapter()
```
Perhaps this is a neat way of solving it, I just want this checkpointer lifecycle instance to be added during startup, not during calling stop().

davidegrohmann replied:
Yes, unfortunately doing it that way you suggest, it is very hard to guarantee it that the check pointer is "shutdown" first. This way it might not elegant but solve the mentioned problem and touches the least amount of code since I am doing this change in a maintenance branch.

tinwelint replied:
By adding it to life "last" it should be shutdown "first", that order is reversed during stop/shutdown isn't it?

davidegrohmann replied:
yes it is correct, adding it last means shutting down first

MishaDemianenko commented on community/kernel/src/test/java/org/neo4j/kernel/NeoStoreDataSourceTest.java:
```diff
@@ -242,6 +245,34 @@ public void logModuleSetUpError() throws Exception
     }
 
+    @Test
+    public void shouldAlwaysShutdownLifeEvenWhenCheckPointingFails() throws Exception
```
Is it possible to have also integration test for this?

davidegrohmann replied:
This is already failing in 2 stress tests.

MishaDemianenko replied:
that does not remove the need in IT right?

davidegrohmann replied:
Having 3 tests showing the problem is enough I think.

davidegrohmann replied:
I have added an IT for this scenario

MishaDemianenko replied:
thank you, appreciate!

tinwelint commented:
Btw, what's was the reason for checkpointing failing?

davidegrohmann commented:
@tinwelint I don't know yet, what comes out is a failure in closing the page cache (since we didn't close the store) but the last thing I see in the `debug.log` is a tentative to check point on shutdown.

davidegrohmann committed:
Fix lifecycle bug when stopping NeoDataSource

If the checkpointing step fails in NeoDataSource.stop(), we never call
life.shutdown(). Hence we'll never unmap the store files from the page
cache. So later when we try to close the page cache it will fail due
to still mapped files.  This change we'll make sure that we always
shutdown all the lifecycled instances in the internal life in
NeoDataSource.

davidegrohmann committed:
Register the checkpoint trigger during life creation

tinwelint commented:
@davidegrohmann This is a slightly scary change in 3.0 since it changes behaviour of shutdown in the event of failure during shutdown, a failure which we don't know what it is. If this is a response to some of our own test failures I think it should not go to maintenance until we (a) understand the problem and (b) potentially have users in real scenarios experiencing this. Otherwise risk is that the change itself breaks something else, isn't it?

davidegrohmann commented:
@tinwelint you have 2 tests in the code base showing the bug! how is that not enough?
